File: src/main/java/blog/typed/javadsl/Buncher.java
package blog.typed.javadsl;

import java.util.ArrayList;
import java.util.Collections;
import java.util.List;
import scala.concurrent.duration.FiniteDuration;

import akka.actor.typed.ActorRef;
import akka.actor.typed.Behavior;
import akka.actor.typed.javadsl.Actor;
import akka.actor.typed.javadsl.TimerScheduler;

public abstract class Buncher {
  private Buncher() {
  }

  interface Msg {
  }

  public static final class Batch {
    private final List<Msg> messages;

    public Batch(List<Msg> messages) {
      this.messages = Collections.unmodifiableList(messages);
    }

    public List<Msg> getMessages() {
      return messages;
    }
  }

  private static final Object TIMER_KEY = new Object();

  private static class Timeout implements Msg {
  }

  public static Behavior<Msg> behavior(ActorRef<Batch> target, FiniteDuration after, int maxSize) {
    return Actor.withTimers(timers -> idle(timers, target, after, maxSize));
  }

  private static Behavior<Msg> idle(TimerScheduler<Msg> timers, ActorRef<Batch> target,
      FiniteDuration after, int maxSize) {
    return Actor.immutable(Msg.class)
      .onMessage(Msg.class, (ctx, msg) -> {
        timers.startSingleTimer(TIMER_KEY, new Timeout(), after);
        List<Msg> buffer = new ArrayList<>();
        buffer.add(msg);
        return active(buffer, timers, target, after, maxSize);
      })
      .build();
  }

  private static Behavior<Msg> active(List<Msg> buffer, TimerScheduler<Msg> timers,
      ActorRef<Batch> target, FiniteDuration after, int maxSize) {
    return Actor.immutable(Msg.class)
      .onMessage(Timeout.class, (ctx, msg) -> {
        target.tell(new Batch(buffer));
        return idle(timers, target, after, maxSize);
      })
      .onMessage(Msg.class, (ctx, msg) -> {
        buffer.add(msg);
        if (buffer.size() == maxSize) {
          timers.cancel(TIMER_KEY);
          target.tell(new Batch(buffer));
          return idle(timers, target, after, maxSize);
        } else {
          return active(buffer, timers, target, after, maxSize);
        }
      })
      .build();
    }

}

File: src/main/scala/blog/typed/scaladsl/Buncher.scala
package blog.typed.scaladsl

import scala.concurrent.duration._
import scala.io.StdIn

import akka.actor.typed.ActorRef
import akka.actor.typed.ActorSystem
import akka.actor.typed.Behavior
import akka.actor.typed.scaladsl.Actor
import akka.actor.typed.scaladsl.TimerScheduler

object Buncher {
  trait Msg
  final case class Batch(messages: Vector[Msg])

  private case object TimerKey
  private case object Timeout extends Msg

  def behavior(target: ActorRef[Batch], after: FiniteDuration, maxSize: Int): Behavior[Msg] =
    Actor.withTimers(timers => idle(timers, target, after, maxSize))

  private def idle(timers: TimerScheduler[Msg], target: ActorRef[Batch],
    after: FiniteDuration, maxSize: Int): Behavior[Msg] = {
    Actor.immutable[Msg] { (ctx, msg) =>
      timers.startSingleTimer(TimerKey, Timeout, after)
      active(Vector(msg), timers, target, after, maxSize)
    }
  }

  private def active(buffer: Vector[Msg], timers: TimerScheduler[Msg],
    target: ActorRef[Batch], after: FiniteDuration, maxSize: Int): Behavior[Msg] = {
    Actor.immutable[Msg] { (ctx, msg) =>
      msg match {
        case Timeout =>
          target ! Batch(buffer)
          idle(timers, target, after, maxSize)
        case msg =>
          val newBuffer = buffer :+ msg
          if (newBuffer.size == maxSize) {
            timers.cancel(TimerKey)
            target ! Batch(newBuffer)
            idle(timers, target, after, maxSize)
          } else
            active(newBuffer, timers, target, after, maxSize)
      }
    }
  }
}

object BuncherDestination {

  final case class Info(s: String) extends Buncher.Msg

  val behavior: Behavior[Buncher.Batch] =
    Actor.immutable[Buncher.Batch] { (ctx, batch) =>
      println(s"Got batch of ${batch.messages.size} messages: ${batch.messages.mkString(", ")} ")
      Actor.same
    }
}

object BuncherApp {

  def main(args: Array[String]): Unit = {
    val root = Actor.deferred[Nothing] { ctx =>
      val destination = ctx.spawn(BuncherDestination.behavior, "destination")
      val buncher = ctx.spawn(
        Buncher.behavior(destination, 3.seconds, 10),
        "buncher")

      (1 to 15).foreach { n =>
        buncher ! BuncherDestination.Info(n.toString)
      }

      ctx.schedule(1.second, buncher, BuncherDestination.Info("16"))
      ctx.schedule(2.seconds, buncher, BuncherDestination.Info("17"))

      ctx.schedule(4.seconds, buncher, BuncherDestination.Info("18"))

      Actor.empty
    }
    val system = ActorSystem[Nothing](root, "Sys")
    try {
      // Exit the system after ENTER is pressed
      StdIn.readLine()
    } finally {
      system.terminate()
    }
  }

}
